File: 0347_Top_K_Frequent_Elements/0347_myans.cpp
class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        unordered_map<int, int> frequency;

        for (int num : nums)
        {
            frequency[num]++;
        }

        auto cmp = [&](const pair<int, int> &a, const pair<int, int> &b)
        {
            return a.second < b.second;
        };

        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> pq(cmp);

        for (auto &entry : frequency)
        {
            pq.push(entry);
        }

        vector<int> result;
        for (int i = 0; i < k; i++)
        {
            result.push_back(pq.top().first);
            pq.pop();
        }

        return result;
    }
};

File: 0347_Top_K_Frequent_Elements/0347.myans.rs
use std::collections::{BinaryHeap, HashMap};
use std::cmp::Reverse;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut freq = HashMap::new();

        for num in nums {
            *freq.entry(num).or_insert(0) += 1;
        }

        let mut heap = BinaryHeap::new();
        for (&num, &count) in freq.iter() {
            heap.push(Reverse((count, num)));
            if heap.len() > k as usize {
                heap.pop();
            }
        }

        heap.into_iter().map(|Reverse((_, num))| num).collect()
    }
}